web/modules/custom/dbcdk_community/src/Form/ProfileEditForm.php

```diff
@@ -222,23 +297,61 @@ public function submitForm(array &$form, FormStateInterface $form_state) {
         // to set the profiles ID to the body of the PUT request.
         $new_data['id'] = $this->profile->getId();
         $this->profile = $this->profileApi->profileUpsert(json_encode($new_data));
-        drupal_set_message($this->t('The profile "%profile" has been updated.', ['%profile' => $this->profile->getUsername()]));
       }
       catch (ApiException $e) {
         \Drupal::logger('DBCDK Community Service')->error($e);
       }
+      return TRUE;
     }
     else {
-      drupal_set_message(
-        $this->t('There were no changes detected for the profile "%profile".', ['%profile' => $this->profile->getUsername()]),
-        'warning'
-      );
+      return FALSE;
     }
+  }
 
-    // Redirect to the Profiles display page.
-    $form_state->setRedirect('page_manager.page_view_dbcdk_community_profile', [
-      'username' => $this->profile->getUsername(),
-    ]);
+  /**
+   * Update a Community Profile's roles.
+   *
+   * Find out if we have to remove or add any roles from a Community Service
+   * based on values from $form_state and the current Community Profiles roles
+   * and return a boolean representing if anything were updated or not.
+   *
+   * @param array $roles
+   *   The value of the roles input from form_state.
+   *
+   * @return bool
+   *   Returns TRUE if there were any changes to update and vise versa.
+   */
+  protected function updateProfileRoles(array $roles) {
+    // Format an array of the profiles role ids before any update has happened.
+    $old_roles = array_map(function($role) {
+      return (int) $role->getId();
+    }, $this->profile->communityRoles);
+
+    // Format an array of the role ids we want the profile to have.
+    $new_roles = array_filter($roles);
+
+    // We find the difference from the new roles and the old once to see if any
+    // roles have to be added to the community profile.
+    if ($add_roles = array_diff($new_roles, $old_roles)) {
+      foreach ($add_roles as $role) {
+        $this->profileApi->profilePrototypeLinkCommunityRoles((int) $role, $this->profile->getId());
+      }
+    };
+
+    // We find the difference from the old roles and the new once to see if any
+    // roles have to be removed from the community profile.
+    if ($remove_roles = array_diff($old_roles, $new_roles)) {
+      foreach ($remove_roles as $role) {
+        $this->profileApi->profilePrototypeUnlinkCommunityRoles((int) $role, $this->profile->getId());
```

Comment:
This may cause an ApiException to be thrown. I do not think there is much use catching it here. In that case we should add a @throws ApiException to the DocBlock.